File: libs/c3dclassessdk/csystem/cdevice/cmemory/CMemory.java
//----------------------------------------------------------------
// file: CMemory
// desc: 
//----------------------------------------------------------------
package c3dclasses;

//----------------------------------------------------------------
// class: CMemory
// desc: 
//----------------------------------------------------------------
public class CMemory extends CResource { 
	public CMemory() { 
		super(); 
		this.m_cache = null; 
	} // end CMemory()
	
	// open / close
	public boolean open(String strpath, String strtype, CHash params) {
		if(!super.open(strpath, strtype, params))
			return false;	
		if(this.driver("open", _.args(this)) == null)
			return false;	
		this.m_cache = _.chash();
		if(params != null && params.containsKey("cmemory_cache")) {
			this.m_cache.append((CHash)params._("cmemory_cache"));	
			return true;
		} // end if
		CReturn creturn = this.sync();
		if(creturn.isdone() && creturn.data() != null) {
			this.m_cache = creturn.chash();
		} // end if
		return true;	
	} // end open()
	
	public boolean close() {
		CReturn creturn = this.driver("close", _.args(this));		
		this._("cmemory_driver_type", "");
		this.m_cache = null; 
		return true;
	} // end close()	
	
	// remote CRUD
	public CReturn create(String strname, Object value, String strtype, CHash params) {
		CReturn creturn = this.driver("create", _.args(this, strname, value, strtype, params));
		if(creturn.isdone() && creturn.data() != null)
			this.m_cache._(strname,creturn.data());
		return creturn; 
	} // end create()
    
	public CReturn retrieve(String strname) {
		CReturn creturn = this.driver("retrieve", _.args(this, strname));
		if(creturn.isdone() && creturn.data() != null)
			this.m_cache._(strname,creturn.data());
		return creturn;
	} // end retrieve()
	
	public CReturn update(String strname, Object value, String strtype, CHash params) {
		CReturn creturn = this.driver("update",_.args(this, strname, value, strtype, params));
		if(creturn.isdone() && creturn.data() != null)
			this.m_cache._(strname,creturn.data());
		return creturn; 
	} // end update()
	
	public CReturn delete(String strname) {
		CReturn creturn = this.driver("delete", _.args(this, strname));
		if(creturn.isdone() && creturn.data() != null)
			this.m_cache.remove(strname);
		return creturn;
	} // end delete()
	
	public CReturn sync() { 
		CReturn creturn = this.driver("sync", _.args(this));
		if(creturn.isdone() && creturn.data() != null)
			this.cache(creturn.chash());
		return creturn;
	} // end sync()
	
	public CReturn upsert(String strname, Object value, String strtype, CHash params) {
		CReturn creturn = this.create(strname, value, strtype, params);	
		return (creturn == null || creturn.data() == null) ? 
			this.update(strname, value, strtype, params) : creturn;
	} // end create_update()
	
	// set / get
	public void set(String strname, Object value, String strtype) {
		if(this.m_cache._(strname) == null)
			return;
		CHash cvar = this.m_cache._chash(strname);
		cvar._("m_value", value);
		this.update(strname, value, strtype, null);
		return;
	} // end set()
	
	public CHash get(String strname) { 
		return (CHash) this.m_cache._chash(strname); 
	} // end get()
	
	// other
	public CHash cache() { 
		return this.m_cache; 
	} // cache()
	
	public CHash cache(CHash cache) { 
		return this.m_cache = cache; 
	} // end cache()
	
	public String toString() { 
		return super.toJSON(false) + "\n" + this.m_cache.toJSON(false); 
	} // end toString()
	
	public CReturn driver(String func, CArray args) { 
		CFunction cfunction = CFunction.get(this._string("cmemory_driver_type") + "." + func);
		return (cfunction != null) ? cfunction.call(args) : CReturn._error(null);		
	} // end driver()
	
	// include / use 
	static public CResource include(String strid, String strdriverpath, String strdrivertype, CHash params) {
		return (CDriver.include(strdrivertype)) ? CResource.include(strid, strdriverpath, 				
				"c3dclasses.CMemory", _.chash(params)._("cmemory_driver_type", strdrivertype)) : null;
	} // end include()
	
	static public CMemory use(String strid) { 
		return (CMemory) CResource.use(strid); 
	} // end use()
	
	protected CHash m_cache = null;
} // end CMemory()

File: libs/_libs/c3dclassessdk/csystem/cdevice/cmemory/CMemory.js
//----------------------------------------------------------------
// file: cmemory.js
// desc: defines the memory object
//----------------------------------------------------------------

//----------------------------------------------------------------
// class: CMemory2
// desc: defines the memory object
//----------------------------------------------------------------
var CMemory2 = new Class({
	Extends : CResource,	// inherits the CResource
	initialize : function() { // CMemory2
		this.parent();
		this.m_cache = {}; 	
		this.m_cmemorydriver = null;
	}, // end CMemory2()
	
	//////////////////////
	// opening / closing
	open : function(strpath, params){
		if(!this.parent(strpath, params) || CMemoryDriver._open(this) == null)
			return false;
		// preload the cache
		this.m_cache = (params["cmemory_cache"]) ? params["cmemory_cache"] : {}; // preload the cache
		return true;
	}, // end open()	
	
	close : function(){
		this.m_cache = {};
		this.m_cmemorydriver = null;
		return CMemoryDriver._close(this);
	}, // end close()

	//////////////////////////
	// async CRUD operations
	create : function(strname, value, strtype, params) {
		return CMemoryDriver._create(this, {
			"m_strname":strname, 
			"m_value":value, 
			"m_strtype":strtype, 
			"m_params": params 
		}); // end CMemoryDriver.create()
	}, // end create()
	
	retrieve : function(strname){ 
		return CMemoryDriver._retrieve(this, strname);
	}, // end retrieve()
	
	update : function(strname, value, strtype, params){
		return CMemoryDriver._update(this, {
			"m_strname":strname, 
			"m_value":value, 
			"m_strtype":strtype, 
			"m_params": params 
		}); // end CMemoryDriver.update()
	}, // end update()
	
	delete : function(strname){
		return CMemoryDriver._delete(this, strname);
	}, // end delete()

	/////////////////////////
	// syncing
	sync : function() {
		return CMemoryDriver._sync(this); // sync local and remote memory
	}, // end sync() 

	////////////////////////
	// other
	cache : function() {
		return this.m_cache; 
	}, // end cache()

	setCMemoryDriver : function(cmemorydriver) {
		this.m_cmemorydriver = cmemorydriver;
	}, // end setCMemoryDriver()

	getCMemoryDriver : function() {
		return this.m_cmemorydriver;
	}, // end getCMemoryDriver()

	_toString : function(){
		return print_r(this.m_cache, true);
	}, // end toString()
	
	/////////////////////////
	// set / get
	set : function(strname, value) {
		if(!this.m_cache || !this.m_cache[strname])
			return;
		var cvar = this.m_cache[strname];
		cvar["m_value"]= value;
		CMemoryDriver._update(this, cvar);
		return;
	}, // end set()
	
	get : function(strname) {
		return (this.m_cache) ? this.m_cache[strname] : null;
	} // end get()
}); // end CMemory2()

/////////////////////////
// includes and using
function include_memory2(strid, strpath, strtype, params) {
	params = params || {};
	params["cresource_type"] = "CMemory2"
	params["cmemorydriver_type"] = strtype;
	params["cmemorydriver_path"] = strpath;
	strpath = "CMemory2||"+strtype+"||"+strpath;
	return include_resource(strid, strpath, params);
} // end include_memory2()

function include_remote_memory2(strid, strpath, strtype, struri, params){
	params = params || {};
	params["cremotememorydriver_path"]=strpath;
	params["cremotememorydriver_type"] = strtype;
	params["cremotememorydriver_uri"] = struri; 	
	params["cremotememorydriver_id"]=struri+"||"+strtype+"||"+strpath;
	return include_memory2(strid, strpath, "CRemoteMemoryDriver", params);
} // end include_remote_memory2()

function use_memory2(strid) {
	return use_resource(strid)
} // end use_memory2()

///////////////////////////////
// importing / exporting
function import_cmemory(strid, params) {
	if(!params)
		return false;
	var strpath = params["cmemorydriver_path"];
	var strtype = params["cmemorydriver_type"];
	var struri = params["cremotememorydriver_uri"];
	return include_remote_memory2(strid, strpath, strtype, struri, params);
} // end import_cmemory()